include/exadg/operators/adaptive_mesh_refinement.h

```diff
@@ -0,0 +1,163 @@
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *  ______________________________________________________________________
+ *
+ * Source: https://github.com/MeltPoolDG/MeltPoolDG
+ * Author: Peter Munch, Magdalena Schreter, TUM, December 2020
+ */
+
+#ifndef INCLUDE_EXADG_OPERATORS_ADAPTIVE_MESH_REFINEMENT_H_
+#define INCLUDE_EXADG_OPERATORS_ADAPTIVE_MESH_REFINEMENT_H_
+
+// deal.II
+#include <deal.II/base/quadrature_lib.h>
+#include <deal.II/distributed/grid_refinement.h>
+#include <deal.II/distributed/solution_transfer.h>
+#include <deal.II/dofs/dof_tools.h>
+#include <deal.II/grid/grid_refinement.h>
+#include <deal.II/numerics/error_estimator.h>
+#include <deal.II/numerics/solution_transfer.h>
+
+
+namespace ExaDG
+{
+struct AdaptiveMeshRefinementData
+{
+  unsigned int trigger_every_n_time_steps = 1;
+
+  int  maximum_refinement_level = 10;
+  int  minimum_refinement_level = 0;
+  bool preserve_boundary_cells  = false;
+
+  double fraction_of_cells_to_be_refined   = 0.0;
+  double fraction_of_cells_to_be_coarsened = 0.0;
+};
+
+/**
+ * Check if adaptive mesh refinement should be triggered depending on the time_step_number.
+ */
+inline bool
+trigger_coarsening_and_refinement_now(const AdaptiveMeshRefinementData & amr_data,
+                                      const unsigned int                 time_step_number)
+{
+  return ((time_step_number == 0) or not(time_step_number % amr_data.trigger_every_n_time_steps));
+}
+
+/**
+ * Limit the coarsening and refinement by adapting the flags set in the triangulation.
+ * This considers the maximum and minimum refinement levels provided and might remove
+ * refinement flags set on the boundary if requested.
+ */
+template<int dim>
+void
+limit_coarsening_and_refinement(dealii::Triangulation<dim> &       tria,
+                                AdaptiveMeshRefinementData const & amr_data)
+{
+  for(auto & cell : tria.active_cell_iterators())
+  {
+    if(cell->is_locally_owned())
+    {
+      // Clear refinement flags on maximum refinement level.
+      if(cell->level() == amr_data.maximum_refinement_level)
+      {
+        cell->clear_refine_flag();
+      }
+
+      // Clear coarsening flags on minimum refinement level.
+      if(cell->level() <= amr_data.minimum_refinement_level)
+      {
+        cell->clear_coarsen_flag();
+      }
+
+      // Do not coarsen/refine cells along boundary if requested
+      if(amr_data.preserve_boundary_cells and cell->at_boundary())
+      {
+        cell->clear_refine_flag();
+        cell->clear_coarsen_flag();
+      }
+    }
+  }
+}
+
+/**
+ * Loop over the locally owned cells and check for refinement/coarsening flags set.
```

Comment:
Involves an MPI call, so I am not sure if the documentation of this function is correct. The comment could maybe appear inside the function.